src/ucs/datastruct/callbackq.c

```diff
@@ -41,354 +49,342 @@ typedef struct ucs_callbackq_oneshot_elem {
+UCS_ARRAY_DEFINE_INLINE(ucs_callbackq_idx, int, unsigned);
+UCS_ARRAY_DEFINE_INLINE(ucs_callbackq_spill_elems, unsigned,
+                        ucs_callbackq_spill_elem_t);
 
-    ucs_callbackq_elem_t     *slow_elems;    /**< Array of slow-path elements */
-    unsigned                 num_slow_elems; /**< Number of slow-path elements */
-    unsigned                 max_slow_elems; /**< Maximal number of slow-path elements */
-    int                      slow_proxy_id;  /**< ID of slow-path proxy in fast-path array.
-                                                  keep track while this moves around. */
+struct ucs_callbackq_priv {
+    /* Protects adding / removing */
+    ucs_recursive_spinlock_t               lock;
 
-    uint64_t                 fast_remove_mask; /**< Mask of which fast-path elements
-                                                    should be removed */
-    unsigned                 num_fast_elems; /**< Number of fast-path elements */
+    /* IDs of fast-path callbacks */
+    int                                    fast_ids[UCS_CALLBACKQ_FAST_COUNT];
 
-    /** Hash map of oneshot path elements, by key */
-    khash_t(ucs_callbackq_oneshot_elems) oneshot_elems;
+    /* Number of fast-path elements */
+    unsigned                               num_fast_elems;
 
-    /* Lookup table for callback IDs. This allows moving callbacks around in
-     * the arrays, while the user can always use a single ID to remove the
-     * callback in O(1).
-     */
-    int                      free_idx_id;    /**< Index of first free item in the list */
-    int                      num_idxs;       /**< Size of idxs array */
-    unsigned                 *idxs;          /**< ID-to-index lookup */
+    /* Mask of which fast-path elements should be removed */
+    uint64_t                               fast_remove_mask;
 
-} ucs_callbackq_priv_t;
+    /* Callback ID to index mapping */
+    ucs_array_t(ucs_callbackq_idx)         idxs;
 
+    /* Index of first free item in the freelist */
+    int                                    free_idx_id;
 
-static unsigned ucs_callbackq_slow_proxy(void *arg);
+    ucs_array_t(ucs_callbackq_spill_elems) spill_elems;
+
+    /* Hash map of oneshot path elements, by key */
+    khash_t(ucs_callbackq_oneshot_elems)   oneshot_elems;
+
+    /* ID of oneshot-path proxy in fast-path array */
+    int                                    proxy_cb_id;
+};
 
-static inline ucs_callbackq_priv_t* ucs_callbackq_priv(ucs_callbackq_t *cbq)
-{
-    UCS_STATIC_ASSERT(sizeof(cbq->priv) == sizeof(ucs_callbackq_priv_t));
-    return (void*)cbq->priv;
-}
+
+static void ucs_callbackq_proxy_enable(ucs_callbackq_t *cbq);
+static void ucs_callbackq_proxy_disable(ucs_callbackq_t *cbq);
 
 static void ucs_callbackq_enter(ucs_callbackq_t *cbq)
 {
-    ucs_recursive_spin_lock(&ucs_callbackq_priv(cbq)->lock);
+    ucs_recursive_spin_lock(&cbq->priv->lock);
+    ucs_log_indent_level(UCS_LOG_LEVEL_TRACE_FUNC, +1);
```

Comment:
As I see we use just `1` in case of the indentation increasing, so `+` can be removed here.